File: sorting/01_merge_sorted_arrays/merge_sorted_arrays.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <cstdint>
#include <vector>

using namespace std;


class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        if (!checkConstraints(nums1, m, nums2, n)) {
            throw std::runtime_error("Constraints violated");
        }
        int p1 = m - 1;
        int p2 = n - 1;
        for (int p = m + n - 1; p > -1; p--) {
            if (p2 < 0) break;
            if ((p1 >= 0) && (nums1[p1] > nums2[p2])) {
                nums1[p] = nums1[p1];
                p1--;
            } else {
                nums1[p] = nums2[p2];
                p2--;
            }
        }
    }

private:
    bool checkConstraints(vector<int> nums1, int m, vector<int> nums2, int n) {
        if (nums1.size() != m + n) return false;
        if (nums2.size() != n) return false;
        if (n < 0 || n > 200 || m < 0 || m > 200) return false;
        if ((m + n) < 1 || (m + n) > 200) return false;
        for (int i = 0; i < m; i++) {
            if (nums1[i] < -pow(10, 9) || nums1[i] > pow(10, 9)) return false;
        }
        for (int j = 0; j < n; j++) {
            if (nums2[j] < -pow(10, 9) || nums2[j] > pow(10, 9)) return false;
        }
        return true;
    }
};

File: sorting/01_merge_sorted_arrays/merge_sorted_arrays.py
from typing import List


class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        if not self._check_constraints(nums1, m, nums2, n):
            raise ValueError('Constraints violated')
        p1 = m - 1
        p2 = n - 1
        for p in range(m + n - 1, -1, -1):
            if p2 < 0:
                break
            if p1 > -1 and nums1[p1] > nums2[p2]:
                nums1[p] = nums1[p1]
                p1 -= 1
            else:
                nums1[p] = nums2[p2]
                p2 -= 1

    def _check_constraints(self, nums1: List[int], m: int, nums2: List[int], n: int) -> bool:
        if len(nums1) != m + n:
            return False
        if len(nums2) != n:
            return False
        if n < 0 or n > 200 or m < 0 or m > 200:
            return False
        if (m + n) < 1 or (m + n) > 200:
            return False
        for i in range(m):
            if nums1[i] < -10**9 or nums1[i] > 10**9:
                return False
        for j in range(n):
            if nums2[j] < -10**9 or nums2[j] > 10**9:
                return False
        return True
